src/OrleansRuntime/MembershipService/MembershipTableFactory.cs

```diff
@@ -28,44 +28,31 @@ public MembershipTableFactory(IServiceProvider serviceProvider, LoggerWrapper<Me
             using (await this.initializationLock.LockAsync())
             {
                 if (membershipTable != null) return membershipTable;
-
-                var globalConfig = this.serviceProvider.GetRequiredService<GlobalConfiguration>();
-                IMembershipTable result;
-                switch (globalConfig.LivenessType)
+                
+                // get membership through DI
+                var result = this.serviceProvider.GetService<IMembershipTable>();
+                //if empty, then try to check if user configured using GranBasedMembership
+                if (result == null)
                 {
-                    case LivenessProviderType.MembershipTableGrain:
+                    //if configured through UseGrainBasedMembershipTable method on ISiloHostBuilder, then this won't be null
+                    var options = this.serviceProvider.GetService<GrainBasedMembershipTableOptions>();
```

Comment:
MembershipTableGrain should be a treated like the others and instead of using dummy GrainBasedMembershipTableOptions class for check, we can register that by default during the setup phase of the silo like MVC is registering default implementations. The check based on the GlobalConfiguration liveness provider type value should not have this logic in MembershipTableFactory, the Legacy Config transform extension method would be a better place if we really need to keep it.
Long term SiloBuilder should be able to check before building if an IMembershipTable is configured for the silo or not.

@jdom, thoughts?

Reply:
Agree, all of this initialization should happen as part of the `IMembershipTable.InitializeMembershipTable()` of the concrete implementation. It does require several changes, so I'd be OK if we postpone that refactoring in a separate PR.
What is confusing is how come you get the `GrainBasedMembershipTableOptions` directly from the container. What's in the container is typically the `IOptions<GrainBasedMembershipTableOptions>`

---
In reply to: [142585221](https://github.com/dotnet/orleans/pull/3490#discussion_r142585221) [](ancestors = 142585221)

Reply:
I vote to refactor it in another PR. There's enough to review in this PR. 
And yes I think you are right about the IOptions<GrainBasedMembershipTableOptions>